{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6a11c566-dfa6-481e-b57f-228efcf08c54",
   "metadata": {},
   "source": [
    "### Colombia 012025 sample processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1308582c-877d-47ca-993c-6dbcaad36db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "ws_allocate col_data 30\n",
    "# /scratch4/workspace/nikea_ulrich_uml_edu-col_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaf34a8c-69ac-4bcb-aac1-786d38af35c3",
   "metadata": {},
   "source": [
    "Ran fastqc/multiqc on raw reads \\\n",
    "**Reads with high adaptor content**\n",
    "- 012025_COL_SAN_T5_569_DLAB_S2 R1 & R2\n",
    "- 012025_COL_SAN_T5_571_DLAB_S4 R1 & R2\n",
    "- 012025_COL_SAN_T5_573_PSTR_S6 R1 & R2\n",
    "- 012025_COL_SAN_T5_582_PSTR_S15 R1 & R2\n",
    "- 012025_COL_SAN_T5_586_DLAB_S19 R1 & R2\n",
    "- 012025_COL_SAN_T5_590_PSTR_S23 R1 & R2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73cd61ed-b0e3-41b3-b356-686518b2ea89",
   "metadata": {},
   "source": [
    "### QC - trim galore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e97de7ec-c3f6-4147-b716-71f58814b128",
   "metadata": {},
   "outputs": [],
   "source": [
    "#make sample id list\n",
    "ls *R1_001.fastq.gz > 012025_sampleids.txt\n",
    "#then remove _R1_001.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "741d9a3d-8592-4233-8713-f04e1a42cc08",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=100G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 48:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-012025qc-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "\n",
    "# Run qc with trim galore and fastqc\n",
    "conda activate qc\n",
    "\n",
    "# Define the paths and variables\n",
    "FILEPATH=\"/scratch3/workspace/nikea_ulrich_uml_edu-data_download/012025_COL_SAN_T5_raw\"\n",
    "OUTDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/trimmed\"\n",
    "mkdir -p $OUTDIR\n",
    "NSLOTS=4  \n",
    "\n",
    "SAMPLE_NAMES_FILE=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/012025_sampleids.txt\"\n",
    "\n",
    "# Check if the file exists\n",
    "if [ ! -e \"$SAMPLE_NAMES_FILE\" ]; then\n",
    "    echo \"Error: $SAMPLE_NAMES_FILE does not exist.\"\n",
    "    exit 1\n",
    "fi\n",
    "\n",
    "# Read each line from the file and perform actions\n",
    "while IFS= read -r sample_id; do\n",
    "    # Form the full file names\n",
    "    input_r1=\"$FILEPATH/${sample_id}_R1_001.fastq.gz\"\n",
    "    input_r2=\"$FILEPATH/${sample_id}_R2_001.fastq.gz\"\n",
    "    \n",
    "    # Ensure the input files exist before running the tools\n",
    "    if [ ! -e \"$input_r1\" ] || [ ! -e \"$input_r2\" ]; then\n",
    "        echo \"Error: Input files do not exist for sample $sample_id\"\n",
    "        continue\n",
    "    fi\n",
    "\n",
    "    # Run trim_galore\n",
    "    trim_galore -j \"$NSLOTS\" -q 20 --phred33 --length 20 --paired $input_r1 $input_r2 --fastqc -o $OUTDIR\n",
    "\n",
    "done < \"$SAMPLE_NAMES_FILE\"\n",
    "\n",
    "\n",
    "# JOB-ID: 49466490\n",
    "# bash script: COL_qc.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bddd5c24-07c3-43ef-bc0b-aa35c2d2e53f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run multiqc\n",
    "module load conda/latest\n",
    "conda activate multiqc\n",
    "multiqc ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c68882a-4c9b-4d09-b2bf-2cbde00adf63",
   "metadata": {},
   "source": [
    "All reads failed the per base sequence content but see https://sequencing.qcfail.com/articles/positional-sequence-bias-in-random-primed-libraries/. Shouldn't be a problem, has more to do with the library prep kit that does random PCR. Proceeding to next step"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a630e066-d9e0-478b-ad52-14961876f8a3",
   "metadata": {},
   "source": [
    "### Host removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed267f9d-9110-493d-aed4-50bf43e852bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 48:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-mcav_hostremoval-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate anvio-8\n",
    "\n",
    "#Remove host from sample reads\n",
    "#set general parameters:\n",
    "SAMPLENAME=\"mcav\"\n",
    "SAMPLELIST=\"012025_${SAMPLENAME}_sampleids.txt\" \n",
    "#manually created the mcav_sampleids file from the whole sample list\n",
    "RAWREADSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/trimmed\"\n",
    "READSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_host_removed/${SAMPLENAME}\"\n",
    "mkdir -p $READSPATH\n",
    "EXTRAFILESPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_host_removed/${SAMPLENAME}/bams\"\n",
    "mkdir -p $EXTRAFILESPATH\n",
    "LISTPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL\"\n",
    "#set step parameters \n",
    "GENOME=\"Mcav\"\n",
    "INPUTPATH=\"/project/pi_sarah_gignouxwolfsohn_uml_edu/Reference_genomes/${GENOME}_genome\"\n",
    "INDEX=\"${GENOME}_DB\"\n",
    "\n",
    "#loop through samples\n",
    "while IFS= read -r SAMPLEID; do\n",
    "#re-align reads back to the index\n",
    "bowtie2 -p 8 -x $INPUTPATH/$INDEX -1 $RAWREADSPATH/\"${SAMPLEID}_R1_001_val_1.fq.gz\" -2 $RAWREADSPATH/\"${SAMPLEID}_R2_001_val_2.fq.gz\" -S $EXTRAFILESPATH/\"${SAMPLEID}\"_mapped_and_unmapped.sam\n",
    "#convert sam file from bowtie to a bam file for processing\n",
    "samtools view -bS $EXTRAFILESPATH/\"${SAMPLEID}\"_mapped_and_unmapped.sam > $EXTRAFILESPATH/\"${SAMPLEID}\"_mapped_and_unmapped.bam\n",
    "#extract only the reads of which both do not match against the host genome\n",
    "samtools view -b -f 12 -F 256 $EXTRAFILESPATH/\"${SAMPLEID}\"_mapped_and_unmapped.bam > $EXTRAFILESPATH/\"${SAMPLEID}\"_bothReadsUnmapped.bam\n",
    "# sorts the file so both mates are together and then extracts them back as .fastq.gz files\n",
    "samtools sort -n -m 5G -@ 2 $EXTRAFILESPATH/\"${SAMPLEID}\"_bothReadsUnmapped.bam -o $EXTRAFILESPATH/\"${SAMPLEID}\"_bothReadsUnmapped_sorted.bam\n",
    "samtools fastq -c 6 -@ 8 $EXTRAFILESPATH/\"${SAMPLEID}\"_bothReadsUnmapped_sorted.bam \\\n",
    "    -1 $READSPATH/\"${SAMPLEID}\"_host_removed_R1.fastq.gz \\\n",
    "    -2 $READSPATH/\"${SAMPLEID}\"_host_removed_R2.fastq.gz \\\n",
    "    -0 /dev/null -s /dev/null -n\n",
    " if [ $? -eq 0 ]; then\n",
    "        echo \"host removal completed successfully for sample: $SAMPLEID\"\n",
    "    else\n",
    "        echo \"host removal encountered an error for sample: $SAMPLEID\"\n",
    "        exit 1  \n",
    "    fi\n",
    "done < \"$LISTPATH/${SAMPLELIST}\"\n",
    "conda deactivate\n",
    "echo \"Host removal: All samples processed successfully.\"\n",
    "\n",
    "# JOB-ID: 49505851\n",
    "# bash script file name: Col_host_removal_mcav.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aabcab48-dd98-4b06-b93a-2d779416e1b3",
   "metadata": {},
   "source": [
    "ran the other species. pstr used Cnat reference, all others used reference genomes of the same species \\\n",
    "ofav: 49505852 \\\n",
    "pstr: 49505853 \\\n",
    "dlab: 49505956 (032024 samples were mapped to Cnat, so need to re-process those by mapping to the new draft ref DLAB)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43491f8c-57f2-4b3d-8ef7-6075b3cfee37",
   "metadata": {},
   "source": [
    "### Symbiont screening + removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83e91b3-0286-4063-878f-aed8573342ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 48:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-mcav-symbremoval-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate assembly\n",
    "\n",
    "#Remove symbiont and human seqs using fastq screen \n",
    "SAMPLENAME=\"mcav\"\n",
    "SAMPLELIST=\"012025_${SAMPLENAME}_sampleids.txt\"  \n",
    "READSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_host_removed/${SAMPLENAME}\"\n",
    "LISTPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL\"\n",
    "FASTQSCREEN='/home/nikea_ulrich_uml_edu/.conda/envs/assembly/share/fastq-screen-0.16.0-0'\n",
    "OUTDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_final_filtered\"\n",
    "mkdir -p \"$OUTDIR\"\n",
    "\n",
    "while IFS= read -r SAMPLEID; do\n",
    "$FASTQSCREEN/fastq_screen --nohits --aligner bowtie2 --conf $FASTQSCREEN/fastq_screen.conf --outdir $OUTDIR \\\n",
    "$READSPATH/\"${SAMPLEID}\"_host_removed_R1.fastq.gz $READSPATH/\"${SAMPLEID}\"_host_removed_R2.fastq.gz;\n",
    " if [ $? -eq 0 ]; then\n",
    "        echo \"fastq_screen completed successfully for sample: $SAMPLEID\"\n",
    "    else\n",
    "        echo \"fastq_screen encountered an error for sample: $SAMPLEID\"\n",
    "        exit 1\n",
    "    fi\n",
    "# --nohits = output reads do not map to any genomes\n",
    "done < \"$LISTPATH/${SAMPLELIST}\"\n",
    "conda deactivate\n",
    "echo \"Symbiont removal: All samples processed successfully.\"\n",
    "\n",
    "# JOB-ID: 50003769\n",
    "# bash script file name: Col_symb_removal.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d71e467e-b309-48ab-9c4d-e785ca403799",
   "metadata": {},
   "source": [
    "ran the other species \\\n",
    "dlab: 49608039 \\\n",
    "pstr: 49608608 \\\n",
    "ofav: 49608433"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3205d36-b2ad-4344-8c7d-788327ec3d0a",
   "metadata": {},
   "source": [
    "for some reason host removal wasn't completed for 012025_COL_SAN_T5_595_MCAV_S28 \\\n",
    "so reran host removal: 50157203 \\\n",
    "and then symb removal: 50242334"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e15003a1-543e-456f-9c23-ca86543feb09",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd /scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_final_filtered\n",
    "conda activate mutliqc\n",
    "multiqc ."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a536ef58-0deb-4ad6-87cc-fa1a25b1171f",
   "metadata": {},
   "source": [
    "before concatenating and performing co-assemblies let's try to blast these filtered seqs against a database composed of the probiotics genomes to get an idea about how abundant they are in this MG data. can't blast with fastq files but I could maybe map with bowtie. actually might make sense to assemble first and the blast to that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "449a4134-ac70-4344-b64e-d4af58e885f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "sed \"s/^>/>7062_015_/\" 7062_015_filtered_assembly.fasta > 7062_015_filtered_assembly_renamed.fasta \n",
    "# this adds 7062_015_ to each header (right after >)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37e3f168-9683-40b0-9a78-05d349622db4",
   "metadata": {},
   "source": [
    "#### repair reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b4207bb-4850-4739-9dbf-02b752fd2fbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-repair012025-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate assembly\n",
    "\n",
    "SAMPLELIST=\"012025_sampleids.txt\" \n",
    "READSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_final_filtered\"\n",
    "LISTPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/\"\n",
    "OUTDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_final_filtered/repaired\"\n",
    "mkdir -p \"$OUTDIR\"\n",
    "\n",
    "#Lets try this! Using repair.sh script from:https://jgi.doe.gov/data-and-tools/software-tools/bbtools/bb-tools-user-guide/repair-guide/\n",
    "while IFS= read -r SAMPLEID; do\n",
    "\n",
    "repair.sh in1=$READSPATH/\"${SAMPLEID}\"_host_removed_R1.tagged_filter.fastq.gz in2=$READSPATH/\"${SAMPLEID}\"_host_removed_R2.tagged_filter.fastq.gz \\\n",
    "out1=$OUTDIR/\"${SAMPLEID}\"_host_removed_R1.tagged_filter_ready.fastq.gz out2=$OUTDIR/\"${SAMPLEID}\"_host_removed_R2.tagged_filter_ready.fastq.gz \\\n",
    "outs=$OUTDIR/\"${SAMPLEID}\"singletons.fq repair;\n",
    " if [ $? -eq 0 ]; then\n",
    "        echo \"repair completed successfully for sample: $SAMPLEID\"\n",
    "    else\n",
    "        echo \"repair encountered an error for sample: $SAMPLEID\"\n",
    "        exit 1\n",
    "    fi\n",
    "done < \"$LISTPATH/${SAMPLELIST}\"\n",
    "\n",
    "conda deactivate\n",
    "echo \"Repair: All samples processed successfully.\"\n",
    "\n",
    "# JOB-ID: 50277995\n",
    "# bash script file name: nikea/COL/bash_scripts/bbtools_repair.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
