{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a52584c-2a31-4b38-8d8d-9d04369c3a7a",
   "metadata": {},
   "source": [
    "### Colombia 012025, 032024 combined assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c75ceeef-bc8c-4ba0-b53a-b20950ad3a28",
   "metadata": {},
   "source": [
    "concatenate 012025 reads for each species, removed the re-sequenced 032024 samples (032024_COL_SAN_T5_140_OFAV_S30 and 032024_COL_SAN_T5_163_PSTR_S31) from the 012025 lists, need to concatenate those with the previous samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "360de43d-601a-4191-9a38-268fef50832d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=50G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 5:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-cat-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate assembly\n",
    "\n",
    "SAMPLENAME=\"ofav\"\n",
    "SAMPLELIST=\"012025_${SAMPLENAME}_sampleids.txt\" \n",
    "READSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_final_filtered/repaired\"\n",
    "LISTPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/\"\n",
    "WORKDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/assembly/concat_reads_ready\"\n",
    "\n",
    "# CONCATETATE all f and r seqs into single file (1 for f, 1 for r)\n",
    "# Read the sample IDs from the file\n",
    "while IFS= read -r SAMPLEID; do\n",
    "    # Construct the file paths for forward and reverse reads\n",
    "    FORWARD_READ=\"$READSPATH/${SAMPLEID}_host_removed_R1.tagged_filter_ready.fastq.gz\"\n",
    "    REVERSE_READ=\"$READSPATH/${SAMPLEID}_host_removed_R2.tagged_filter_ready.fastq.gz\"\n",
    "\n",
    "    # Check if the files exist before concatenating\n",
    "    if [ -e \"$FORWARD_READ\" ]; then\n",
    "        cat \"$FORWARD_READ\" >> \"$WORKDIR/012025_${SAMPLENAME}_reads_R1.fastq.gz\"\n",
    "    else\n",
    "        echo \"Forward read file not found for sample $SAMPLEID\"\n",
    "    fi\n",
    "\n",
    "    if [ -e \"$REVERSE_READ\" ]; then\n",
    "        cat \"$REVERSE_READ\" >> \"$WORKDIR/012025_${SAMPLENAME}_reads_R2.fastq.gz\"\n",
    "    else\n",
    "        echo \"Reverse read file not found for sample $SAMPLEID\"\n",
    "    fi\n",
    "done < \"$LISTPATH/${SAMPLELIST}\"\n",
    "\n",
    "conda deactivate\n",
    "echo \"Concatenation completed!\"\n",
    "\n",
    "# JOB-ID:50346545\n",
    "# bash script file name: nikea/COL/bash_scripts/Col_concat.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57787e9d-baab-4a68-98ad-700a34aab878",
   "metadata": {},
   "source": [
    "IDs for the other species \\\n",
    "pstr: 50347739 \\\n",
    "dlab: 50347756 \\\n",
    "mcav: 50347784"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6cb4697-51cb-4f79-81a2-ca8207fe5e17",
   "metadata": {},
   "source": [
    "concatenated the re-sequenced samples to original ones and re-concatenated the 032024 to include them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f204238d-3ea2-4e0e-8842-332ab6fe818b",
   "metadata": {},
   "source": [
    "concatenated \\\n",
    "032024_COL_SAN_T5_140_OFAV_S30_host_removed_R1.tagged_filter_ready.fastq.gz and 032024_COL_SAN_T5_140_OFAV_S18_host_removed_R1.tagged_filter_ready.fastq.gz \\\n",
    "032024_COL_SAN_T5_140_OFAV_S30_host_removed_R2.tagged_filter_ready.fastq.gz and 032024_COL_SAN_T5_140_OFAV_S18_host_removed_R2.tagged_filter_ready.fastq.gz \\\n",
    "\n",
    "032024_COL_SAN_T5_163_PSTR_S31_host_removed_R1.tagged_filter_ready.fastq.gz and 032024_COL_SAN_T5_163_PSTR_S11_host_removed_R1.tagged_filter_ready.fastq.gz \\\n",
    "032024_COL_SAN_T5_163_PSTR_S31_host_removed_R2.tagged_filter_ready.fastq.gz and 032024_COL_SAN_T5_163_PSTR_S11_host_removed_R2.tagged_filter_ready.fastq.gz \\ \n",
    "\n",
    "new files in /032024_resequenced \\\n",
    "032024_COL_SAN_T5_140_OFAV_S18c R1 and R2 \\\n",
    "032024_COL_SAN_T5_163_PSTR_S11c R1 and R2 \\\n",
    "**moved these samples back into species folders with the other 032024 samples, and reconcatenated**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cfe0052-6351-4216-bf08-3979b8e7fb73",
   "metadata": {},
   "source": [
    "concatenated all samples for each species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f6814cd-9047-4663-85c1-38af52e4df76",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=50G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/assembly/pstr/slurm-cat-%j.out  # %j = job ID\n",
    "\n",
    "SAMPLENAME=\"dlab\"\n",
    "READSPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/assembly/concat_reads_ready\"\n",
    "OUTDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/all_sequences_concat\"\n",
    "\n",
    "cd $READSPATH\n",
    "\n",
    "cat ./\"032024_${SAMPLENAME}_reads_R1.fastq.gz\" ./\"012025_${SAMPLENAME}_reads_R1.fastq.gz\" \\\n",
    ">> $OUTDIR/\"${SAMPLENAME}_all_reads_R1.fastq.gz\"\n",
    "\n",
    "cat ./\"032024_${SAMPLENAME}_reads_R2.fastq.gz\" ./\"012025_${SAMPLENAME}_reads_R2.fastq.gz\" \\\n",
    ">> $OUTDIR/\"${SAMPLENAME}_all_reads_R2.fastq.gz\"\n",
    "\n",
    "\n",
    "# JOB-ID: 50352583\n",
    "# bash script file name: nikea/COL/bash_scripts/Col_concat_all.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06f88017-252a-4301-ba2a-ddef5bb236fe",
   "metadata": {},
   "source": [
    "ran for the others \\\n",
    "mcav: 50353368 \\\n",
    "ofav: 50353621 \\\n",
    "pstr: 50354758"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54a825be-9451-4413-8357-3ea740aa8421",
   "metadata": {},
   "source": [
    "### Megahit assembly all data (both 032024 and 012025)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "114a1cc9-a53b-48ef-9ee3-c88ea2989c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH --qos=long # extend time limit to longer than 2 days\n",
    "#SBATCH -t 72:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-assembly-ofav-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate assembly\n",
    "\n",
    "SAMPLENAME=\"ofav\"\n",
    "WORKDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/all_sequences_concat\"\n",
    "OUTDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/assemblies_all\"\n",
    "mkdir -p $OUTDIR\n",
    "\n",
    "# ASSEMBLE reads into contigs (contiguous sequence - joins them together based on read overlap, and ensures there are no gaps\n",
    "megahit --presets meta-large \\\n",
    "-1 \"$WORKDIR\"/\"$SAMPLENAME\"_all_reads_R1.fastq.gz \\\n",
    "-2 \"$WORKDIR\"/\"$SAMPLENAME\"_all_reads_R2.fastq.gz \\\n",
    "-o $OUTDIR/megahit_assembly_\"$SAMPLENAME\" --out-prefix \"$SAMPLENAME\"_all \\\n",
    "#this one has to make the directory; will fail if it already exists\n",
    "\n",
    "conda deactivate\n",
    "echo \"Assembly completed!\"\n",
    "\n",
    "# JOB-ID: 50356739\n",
    "# bash script file name: nikea/COL/bash_scripts/Col_assemble_ofav.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4409ce92-3406-4d70-9ce4-4ef907bd9097",
   "metadata": {},
   "source": [
    "didn't add \"#SBATCH --qos=long\" to ofav script, but I might need it, so adding it for the others"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd975b5d-60b7-4c69-9a6f-19c60cbd1c00",
   "metadata": {},
   "source": [
    "pstr: 50366433 \\\n",
    "dlab: 50368001 \\\n",
    "mcav: 50367881"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d4f1180-fb25-475f-9ca8-305cd84b859e",
   "metadata": {},
   "source": [
    "I got them running before I realized I didn't add quast into the script ugh. Is it worth it to run metaquast on a co-assembly? "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc4ce24-a820-403f-b41c-c9c595041c61",
   "metadata": {},
   "source": [
    "## Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d10d6592-5e53-4f1d-8f44-490c6b8cfb1b",
   "metadata": {},
   "source": [
    "mapping the co-assemblies that include all data. will do separate scripts for mapping to 012025 data and 032024 because seq files are in different places."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a5997a5-af75-4d98-8e38-02191a883a88",
   "metadata": {},
   "source": [
    "#### 012025 first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e2680ed-d33b-4710-9a16-21de37df7e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH --qos=long # extend time limit to longer than 2 days\n",
    "#SBATCH -t 72:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-mapping012025_ofav-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate anvio-8\n",
    "\n",
    "SAMPLENAME=\"ofav\"\n",
    "READSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/012025_final_filtered/repaired\"\n",
    "CONTIGPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/assemblies_all/megahit_assembly_${SAMPLENAME}\"\n",
    "CONTIGFILE=\"$SAMPLENAME\"_all.contigs.fa\n",
    "WORKPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/012025\"\n",
    "mkdir -p \"$WORKPATH\"\n",
    "XTRAFILES=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/012025/sams\"\n",
    "mkdir -p \"$XTRAFILES\"\n",
    "LISTPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/\"\n",
    "SAMPLELIST=\"012025_${SAMPLENAME}_sampleids.txt\" \n",
    " \n",
    "anvi-script-reformat-fasta $CONTIGPATH/$CONTIGFILE -o $WORKPATH/\"${SAMPLENAME}.contigs-fixed.fsa\" -l 1000 --simplify-names --report-file $WORKPATH/\"${SAMPLENAME}\"_contig-rename-report.txt\n",
    "\n",
    "#fixes deflines (filters contigs and reformats so naming is cleaner)\n",
    "#filtering seq length 1000bp...need to play around with filtering based on bp length\n",
    "#deflines = sequence definition line. comes directly before its associated sequence in a fasta file\n",
    "\n",
    "\n",
    "FIXEDCON=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "\n",
    "cd $WORKPATH\n",
    "#this builds an index of your contigs, which only needs to happen once\n",
    "bowtie2-build $FIXEDCON \"$SAMPLENAME\"_contigs\n",
    "# will not accept path before contigs file - must be in the correct dir \n",
    "\n",
    "while IFS= read -r SAMPLEID; do\n",
    "    #align reads to your contigs and collects that in a .sam file\n",
    "    bowtie2 --threads 11 -x \"$SAMPLENAME\"_contigs -1 $READSPATH/\"${SAMPLEID}\"_host_removed_R1.tagged_filter_ready.fastq.gz -2 $READSPATH/\"${SAMPLEID}\"_host_removed_R2.tagged_filter_ready.fastq.gz -S $XTRAFILES/\"${SAMPLEID}\".sam\n",
    "    #make sure to point it to the index not the FIXEDCON file (-x parameter)\n",
    "    \n",
    "    #converts your sam file to a bam file, but its neither sorted nor indexed, so we use an Anvi'O script to do so:\n",
    "    samtools view -F 4 -b -S $XTRAFILES/\"${SAMPLEID}\".sam -o $WORKPATH/\"${SAMPLEID}\"-RAW.bam\n",
    "   \n",
    "    #index and sort your bam file\n",
    "    anvi-init-bam $WORKPATH/\"${SAMPLEID}\"-RAW.bam -o $WORKPATH/\"${SAMPLEID}\".bam\n",
    "    \n",
    "    rm $WORKPATH/\"${SAMPLEID}\"-RAW.bam\n",
    "done < \"$LISTPATH/${SAMPLELIST}\"\n",
    "echo \"Mapping success!\"\n",
    "\n",
    "#JOB ID: 50420392\n",
    "#bash script: nikea/COL/bash_scripts/Col_mapping_012025_ofav.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b58eacfb-1c96-4c8a-a47a-53e8c38b2a27",
   "metadata": {},
   "source": [
    "job IDs \\\n",
    "mcav: 50427381 \\\n",
    "dlab: 50433768 \\\n",
    "pstr: 50433798"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fa17fae-bc3f-4521-a83f-856ccfbed69c",
   "metadata": {},
   "source": [
    "#### next 032024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff0229cb-6928-4b3b-92de-795a42d80cc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH --qos=long # extend time limit to longer than 2 days\n",
    "#SBATCH -t 72:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-mapping032024_ofav-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate anvio-8\n",
    "\n",
    "SAMPLENAME=\"ofav\"\n",
    "READSPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/assembly/${SAMPLENAME}/repaired\"\n",
    "CONTIGPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/assemblies_all/megahit_assembly_${SAMPLENAME}\"\n",
    "CONTIGFILE=\"$SAMPLENAME\"_all.contigs.fa\n",
    "WORKPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/032024\"\n",
    "mkdir -p \"$WORKPATH\"\n",
    "XTRAFILES=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/032024/sams\"\n",
    "mkdir -p \"$XTRAFILES\"\n",
    "LISTPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/\"\n",
    "SAMPLELIST=\"032024_${SAMPLENAME}_sampleids_new.txt\" \n",
    " \n",
    "#anvi-script-reformat-fasta $CONTIGPATH/$CONTIGFILE -o $WORKPATH/\"${SAMPLENAME}.contigs-fixed.fsa\" -l 1000 --simplify-names --report-file $WORKPATH/\"${SAMPLENAME}\"_contig-rename-report.txt\n",
    "\n",
    "#fixes deflines (filters contigs and reformats so naming is cleaner)\n",
    "#filtering seq length 1000bp...need to play around with filtering based on bp length\n",
    "#deflines = sequence definition line. comes directly before its associated sequence in a fasta file\n",
    "\n",
    "#FIXEDCON=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "CONTIGDB=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/012025\"\n",
    "\n",
    "cd $WORKPATH\n",
    "#this builds an index of your contigs, which only needs to happen once\n",
    "\n",
    "#bowtie2-build $FIXEDCON \"$SAMPLENAME\"_contigs\n",
    "\n",
    "# will not accept path before contigs file - must be in the correct dir \n",
    "\n",
    "while IFS= read -r SAMPLEID; do\n",
    "    #align reads to your contigs and collects that in a .sam file\n",
    "    bowtie2 --threads 11 -x $CONTIGDB/\"$SAMPLENAME\"_contigs -1 $READSPATH/\"${SAMPLEID}\"_host_removed_R1.tagged_filter_ready.fastq.gz -2 $READSPATH/\"${SAMPLEID}\"_host_removed_R2.tagged_filter_ready.fastq.gz -S $XTRAFILES/\"${SAMPLEID}\".sam\n",
    "    #make sure to point it to the index not the FIXEDCON file (-x parameter)\n",
    "    \n",
    "    #converts your sam file to a bam file, but its neither sorted nor indexed, so we use an Anvi'O script to do so:\n",
    "    samtools view -F 4 -b -S $XTRAFILES/\"${SAMPLEID}\".sam -o $WORKPATH/\"${SAMPLEID}\"-RAW.bam\n",
    "   \n",
    "    #index and sort your bam file\n",
    "    anvi-init-bam $WORKPATH/\"${SAMPLEID}\"-RAW.bam -o $WORKPATH/\"${SAMPLEID}\".bam\n",
    "    \n",
    "    rm $WORKPATH/\"${SAMPLEID}\"-RAW.bam\n",
    "done < \"$LISTPATH/${SAMPLELIST}\"\n",
    "echo \"Mapping success!\"\n",
    "\n",
    "#JOB ID: 50427641\n",
    "#bash script: nikea/COL/bash_scripts/Col_032024_mapping_ofav.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5daa5c98-7d48-4041-8723-a4eef1fe9865",
   "metadata": {},
   "source": [
    "job IDs \\\n",
    "mcav: 50434140\\\n",
    "dlab: 50434147\\\n",
    "pstr: 50434158 (used new sample id table here too)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ecddfb-141d-4eae-8a72-799e82aed826",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "322740b7-b5e7-42d0-8d7b-9b5f13818038",
   "metadata": {},
   "source": [
    "#### blasting contigs to probiotics database (custom db I made with the probiotic genomes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "538d6c17-a95c-47ce-afbc-5330ccde42f6",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/books/NBK279684/ good reference for local blast flags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "057d3c01-3948-4cc4-9137-f056ff652ebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=10G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 3:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-probioticsblast-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/.conda/envs/quast #this env has blast\n",
    "\n",
    "#set parameters\n",
    "SAMPLENAME=\"ofav\"\n",
    "PROBPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/cordap_isolates/filtered_assembly/renamed_headers/probiotics_db\"\n",
    "PROBFASTA=\"probiotics7062_1_to_10.fasta\"\n",
    "CONTIGPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/012025\"\n",
    "CONTIGS=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "OUTDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/probiotics_blast\"\n",
    "mkdir -p $OUTDIR\n",
    "\n",
    "#I think it makes sense to make the contig db and then blast the probiotics.fasta to it? or does the other way around make sense?\n",
    "makeblastdb -in $PROBPATH/$PROBFASTA -parse_seqids -dbtype nucl\n",
    "\n",
    "blastn -db $PROBPATH/$PROBFASTA -query $CONTIGPATH/$CONTIGS -num_threads 4 -out $OUTDIR/\"${SAMPLENAME}\"_prob_blast_out.txt \\\n",
    "-outfmt \"6 qseqid sseqid length qstart qend evalue bitscore\" \n",
    "\n",
    "conda deactivate\n",
    "\n",
    "#JOB ID: 50432106\n",
    "#bash script: probiotics_blast_ofav.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e034a54d-9f6b-4c3c-8883-27f2d985bbf7",
   "metadata": {},
   "source": [
    "do for the others, but need to decide if this is helpful or if I need to do another way"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
