{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "16bdaecf-edbf-456d-a7ec-d02e9066c53f",
   "metadata": {},
   "source": [
    "### Colombia 012025, 032024 Binning & Taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b6b87e5-1ba8-4ebf-a158-526ff70c14ac",
   "metadata": {},
   "source": [
    "#### Metabat2, Concoct, Maxbin2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d36636f-e7c4-4155-9fcc-eea2ca869ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=180G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH --qos=long # extend time limit to longer than 2 days if needed\n",
    "#SBATCH -t 72:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-binning-mcav-%j.out  # %j = job ID  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate binning\n",
    "\n",
    "#set parameters for binning:\n",
    "SAMPLENAME=\"mcav\"\n",
    "READSPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/all_sequences_concat\"\n",
    "F_READS=\"${SAMPLENAME}_all_reads_R1.fastq.gz\"\n",
    "R_READS=\"${SAMPLENAME}_all_reads_R2.fastq.gz\"\n",
    "CONTIGPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/assembly/assemblies_all\"\n",
    "CONTIGFILE=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "BAMPATH=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/mapping/all_bams/${SAMPLENAME}\"\n",
    "METABINDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/MetaBAT2_bins\"\n",
    "mkdir -p $METABINDIR\n",
    "MAXBINDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/Maxbin2_bins\"\n",
    "mkdir -p $MAXBINDIR\n",
    "\n",
    "#run MetaBAT2\n",
    "#create depth file for MetaBat2\n",
    "jgi_summarize_bam_contig_depths --outputDepth $METABINDIR/MetaBAT2_depth.txt $BAMPATH/*.bam\n",
    "\n",
    "#MetaBat2 script with verbose output, minimum length (m)(has to be >=1500) and no min bin size \n",
    "metabat2 -i $CONTIGPATH/$CONTIGFILE -a $METABINDIR/MetaBAT2_depth.txt \\\n",
    "-o $METABINDIR/metabat2 -m 1500\n",
    "\n",
    "#run CheckM on metabat2 bins\n",
    "checkm lineage_wf -x fa -t 3 $METABINDIR/ $METABINDIR/checkm-bins-stats\n",
    "\n",
    "conda deactivate\n",
    "echo \"deactivated binning environment\"\n",
    "\n",
    "#run CONCOCT (has it's own environment)\n",
    "module load conda/latest\n",
    "conda activate concoct_env\n",
    "echo \"activated concoct environment\"\n",
    "\n",
    "CONCBINDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/Concoct_bins\"\n",
    "mkdir -p $CONCBINDIR\n",
    "CTEMPDIR=\"/scratch4/workspace/nikea_ulrich_uml_edu-col_data/binning/concoct_${SAMPLENAME}_temp\"\n",
    "mkdir -p $CTEMPDIR\n",
    "\n",
    "#creates the CONCOCT depth file \n",
    "cut_up_fasta.py $CONTIGPATH/$CONTIGFILE -c 10000 -o 0 --merge_last -b $CONCBINDIR/\"${SAMPLENAME}\"_contigs_cut.bed > $CONCBINDIR/\"${SAMPLENAME}\"_contigs_cut.fa\n",
    "\n",
    "#estimate contig coverage\n",
    "concoct_coverage_table.py $CONCBINDIR/\"${SAMPLENAME}\"_contigs_cut.bed $BAMPATH/*.bam > $CONCBINDIR/coverage_table_\"${SAMPLENAME}\".tsv || { echo 'Exit code 2: failed to create coverage file, exiting.' && exit; }\n",
    "\n",
    "#run CONCOCT\n",
    "concoct --composition_file $CONCBINDIR/\"${SAMPLENAME}\"_contigs_cut.fa --coverage_file $CONCBINDIR/coverage_table_\"${SAMPLENAME}\".tsv -t 3 -b $CTEMPDIR || { echo 'Exit code 3: CONCOCT failed to run, exiting.' && exit; }\n",
    "merge_cutup_clustering.py $CTEMPDIR/clustering_gt1000.csv > $CTEMPDIR/\"${SAMPLENAME}\"_clustering_merged.csv || { echo 'Exit code 4: failed to merge clusters, exiting.' && exit; }\n",
    "extract_fasta_bins.py $CONTIGPATH/$CONTIGFILE $CTEMPDIR/\"${SAMPLENAME}\"_clustering_merged.csv --output_path $CONCBINDIR || { echo 'Exit code 5: Bins were not extracted, exiting.' && exit; }\n",
    "\n",
    "conda deactivate\n",
    "echo \"deactivated conconct environment\"\n",
    "\n",
    "conda activate binning\n",
    "echo \"activated binning environment\"\n",
    "\n",
    "#run CheckM on CONCOCT bins\n",
    "checkm lineage_wf -x fa -t 3 $CONCBINDIR/ $CONCBINDIR/checkm-bins-stats\n",
    "\n",
    "#run Maxbin2\n",
    "run_MaxBin.pl -contig $CONTIGPATH/$CONTIGFILE -reads $READSPATH/$F_READS -reads2 $READSPATH/$R_READS -out $MAXBINDIR/maxbin2 -thread 16\n",
    "\n",
    "#run CheckM on Maxbin2 bins\n",
    "checkm lineage_wf -x fasta -t 3 $MAXBINDIR/ $MAXBINDIR/checkm-bins-stats\n",
    "\n",
    "conda deactivate\n",
    "\n",
    "# JOB-ID: 50506413, (t1 script- for troubleshooting steps)  50542337, 50556190\n",
    "# bash script file name: /nikea/COL/bash_scripts/Col_binning.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37626e33-0061-4cf5-99f6-f756b90dedd1",
   "metadata": {},
   "source": [
    "once I got the script running smoothly for mcav samples, submitted it for the others \\\n",
    "dlab: 50556320 \\\n",
    "ofav: 50556315 \\\n",
    "pstr: 50556317 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86c86bf7-0721-41e2-9a5a-7b5b6ac15831",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MetaBAT2 (v2:2.17)\n",
    "# default parameters:\n",
    "#-m [ --minContig ] arg (=2500)    Minimum size of a contig for binning (should be >=1500).\n",
    "#  --maxP arg (=95)                  Percentage of 'good' contigs considered for binning decided by connection\n",
    "#                                    among contigs. The greater, the more sensitive.\n",
    "#  --minS arg (=60)                  Minimum score of a edge for binning (should be between 1 and 99). The \n",
    "#                                    greater, the more specific.\n",
    "#  --maxEdges arg (=200)             Maximum number of edges per node. The greater, the more sensitive.\n",
    "#  --pTNF arg (=0)                   TNF probability cutoff for building TNF graph. Use it to skip the \n",
    "#                                    preparation step. (0: auto).\n",
    "#  -x [ --minCV ] arg (=1)           Minimum mean coverage of a contig in each library for binning.\n",
    "#  --minCVSum arg (=1)               Minimum total effective mean coverage of a contig (sum of depth over \n",
    "#                                    minCV) for binning.\n",
    "#  -s [ --minClsSize ] arg (=200000) Minimum size of a bin as the output.\n",
    "#  -t [ --numThreads ] arg (=0)      Number of threads to use (0: use all cores)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19bdf5bd-5b5a-45d9-a727-b9be74cc4600",
   "metadata": {},
   "source": [
    "resource for maxbin2: https://gtpb.github.io/AM22/pages/07-binning/7.binning.html \\\n",
    "https://nf-co.re/modules/maxbin2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad2b3cdb-0317-443d-b158-46d98075a457",
   "metadata": {},
   "source": [
    "### Das Tool to de-replicate set of bins from multiple binning software"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b81d5d65-65fc-473c-91e2-c9f21af0d476",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prep all the input files for das_tool\n",
    "salloc -c 6 -p cpu\n",
    "#Convert fasta output from MetaBat2, CONCOCT, and Maxbin2 into the correct format for DAS tool \n",
    "conda activate binning\n",
    "Fasta_to_Contig2Bin.sh -i MetaBAT2_bins/ -e fa > ./metabat2.contigs2bin.tsv\n",
    "Fasta_to_Contig2Bin.sh -i Maxbin2_bins/ -e fasta > ./maxbin2.contigs2bin.tsv\n",
    "perl -pe \"s/,/\\tconcoct./g;\" /scratch4/workspace/nikea_ulrich_uml_edu-col_data/binning/concoct_mcav_temp/mcav_clustering_merged.csv > ./concoct.contigs2bin.tsv\n",
    "\n",
    "#remove the first row (heading) of concoct.contigs2bin.tsv\n",
    "sed -i '1d' concoct.contigs2bin.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23991ef4-a259-403d-b8a5-33df44cba5cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=100G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-das_tool-mcav-%j.out  # %j = job ID  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate binning\n",
    "\n",
    "# Set parameters\n",
    "SAMPLENAME=\"mcav\"\n",
    "CONCBINDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/Concoct_bins/concoct.contigs2bin.tsv\"  \n",
    "METABINDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/MetaBAT2_bins/metabat2.contigs2bin.tsv\"\n",
    "MAXBINDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/Maxbin2_bins/maxbin2.contigs2bin.tsv\"\n",
    "CONTIGPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/assembly/assemblies_all\"\n",
    "CONTIGFILE=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "OUTDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/Das_Tool\"\n",
    "mkdir -p $OUTDIR\n",
    "\n",
    "#Run DAS Tool\n",
    "DAS_Tool -i $CONCBINDIR,$METABINDIR,$MAXBINDIR \\\n",
    "-l concoct,metabat2,maxbin2 \\\n",
    "-c $CONTIGPATH/$CONTIGFILE \\\n",
    "-t 11 \\\n",
    "--write_bin_evals \\\n",
    "--write_bins \\\n",
    "-o $OUTDIR/\"${SAMPLENAME}\"\n",
    "\n",
    "#Run CheckM\n",
    "checkm lineage_wf -x fa -t 3 $OUTDIR/\"${SAMPLENAME}\"_DASTool_bins $OUTDIR/CheckM_stats\n",
    "\n",
    "# -i input list: tab seperated table of contigs-bins \n",
    "#--score_threshold default is 0.5\n",
    "\n",
    "# JOB-ID: 50571591\n",
    "# bash script file name: /nikea/COL/bash_scripts/Col_das_tool.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a494816-6438-4ef3-a4dc-31cd5ea89703",
   "metadata": {},
   "source": [
    "other job IDs: \\\n",
    "ofav: 50571677 \\\n",
    "pstr: 50571690 \\\n",
    "dlab: 50571691"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37cd79dd-839b-41cb-b9da-f3a7d2856096",
   "metadata": {},
   "outputs": [],
   "source": [
    "#DLAB\n",
    "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  Bin Id                     Marker lineage            # genomes   # markers   # marker sets    0     1    2    3    4   5+   Completeness   Contamination   Strain heterogeneity  \n",
    "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  metabat2.9         o__Sphingomonadales (UID3310)         26         569           293         5    548   14   2    0   0       99.53            3.04               5.00          \n",
    "  maxbin2.015_sub   c__Alphaproteobacteria (UID3305)      564         349           230         4    334   11   0    0   0       98.91            3.52               0.00          \n",
    "  metabat2.22            k__Bacteria (UID2495)            2993        140            85         1    134   5    0    0   0       98.82            3.64              80.00          \n",
    "  metabat2.7             s__algicola (UID2847)             33         496           263         20   466   10   0    0   0       98.49            1.81              10.00          \n",
    "  metabat2.74            k__Bacteria (UID3187)            2258        188           117         4    182   2    0    0   0       96.58            1.71               0.00          \n",
    "  concoct.75        c__Gammaproteobacteria (UID4267)      119         544           284         21   517   6    0    0   0       94.85            1.06              33.33          \n",
    "  metabat2.43       c__Alphaproteobacteria (UID3305)      564         349           230         28   320   1    0    0   0       90.26            0.43              100.00         \n",
    "  metabat2.54       c__Alphaproteobacteria (UID3305)      564         349           230         31   307   10   1    0   0       88.63            3.68              61.54          \n",
    "  concoct.66_sub       o__Cytophagales (UID2936)           47         454           336         53   378   21   2    0   0       88.40            5.57               3.70          \n",
    "  metabat2.42       c__Alphaproteobacteria (UID3305)      564         349           230         48   292   9    0    0   0       87.40            2.72              44.44          \n",
    "  metabat2.73          p__Cyanobacteria (UID2192)          79         584           458        131   444   9    0    0   0       77.68            1.28              44.44          \n",
    "  concoct.50_sub    c__Alphaproteobacteria (UID3305)      564         349           230         78   199   58   12   2   0       76.00           26.07              81.13          \n",
    "  metabat2.38            k__Bacteria (UID2570)            433         274           183         56   216   2    0    0   0       75.91            0.56               0.00          \n",
    "  metabat2.44           o__Rhizobiales (UID3642)          107         484           316        165   311   8    0    0   0       65.37            1.47              25.00          \n",
    "  concoct.145_sub        k__Bacteria (UID2570)            433         274           183         77   182   13   1    1   0       61.91            6.58              45.45          \n",
    "  concoct.118_sub     o__Actinomycetales (UID1530)        622         259           152        108   138   13   0    0   0       61.56            5.27              23.08          \n",
    "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0efb023a-3ea1-4489-aed6-d464be93b2c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#MCAV\n",
    "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  Bin Id                    Marker lineage            # genomes   # markers   # marker sets    0     1    2    3   4   5+   Completeness   Contamination   Strain heterogeneity  \n",
    "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  metabat2.24         p__Cyanobacteria (UID2192)          79         584           458         9    569   6    0   0   0       98.33            0.84              16.67          \n",
    "  metabat2.3            k__Bacteria (UID3187)            2258        188           117         6    179   3    0   0   0       94.87            2.56              33.33          \n",
    "  concoct.13_sub   c__Alphaproteobacteria (UID3305)      564         349           230         88   236   25   0   0   0       74.12            9.16              44.00          \n",
    "  metabat2.9           o__Rhizobiales (UID3447)          356         416           249        114   298   4    0   0   0       70.73            0.72              25.00          \n",
    "  concoct.58          o__Cytophagales (UID2936)           47         454           336        164   278   12   0   0   0       62.04            2.64              25.00          \n",
    "  metabat2.19           k__Bacteria (UID2570)            433         274           183         84   184   6    0   0   0       57.81            1.36               0.00          \n",
    "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8633557-01d3-43f8-9d92-743c4cc5f227",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OFAV\n",
    "--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  Bin Id                    Marker lineage            # genomes   # markers   # marker sets   0     1    2    3   4   5+   Completeness   Contamination   Strain heterogeneity  \n",
    "--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  concoct.150      c__Gammaproteobacteria (UID4267)      119         544           284        60   467   17   0   0   0       86.32            3.27              64.71          \n",
    "  concoct.61_sub        k__Bacteria (UID2565)            2921        152            93        17   117   18   0   0   0       84.75           11.01              61.11          \n",
    "  metabat2.3        f__Flavobacteriaceae (UID2845)        53         548           298        76   461   10   1   0   0       83.74            2.31              38.46          \n",
    "  concoct.46            k__Bacteria (UID2570)            433         274           183        89   180   5    0   0   0       57.81            1.38              20.00          \n",
    "--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "702ddf58-bd3d-4cef-9b23-6d88e588f224",
   "metadata": {},
   "outputs": [],
   "source": [
    "#PSTR\n",
    "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  Bin Id                    Marker lineage            # genomes   # markers   # marker sets    0     1    2    3   4   5+   Completeness   Contamination   Strain heterogeneity  \n",
    "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "  metabat2.17       o__Sphingomonadales (UID3310)         26         569           293         2    550   15   2   0   0       99.63            3.27               0.00          \n",
    "  concoct.14_sub       o__Rhizobiales (UID3642)          107         484           316         8    469   7    0   0   0       99.30            1.92              28.57          \n",
    "  concoct.18       c__Alphaproteobacteria (UID3305)      564         349           230        111   230   8    0   0   0       64.94            2.16              75.00          \n",
    "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6dc94ae4-2a74-4967-b3df-3c1d94a97c28",
   "metadata": {},
   "source": [
    "### GTDBtk - assigning taxonomy to bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62dad5e3-b332-4a19-a1c8-1aed71845865",
   "metadata": {},
   "outputs": [],
   "source": [
    "#INSTALLATION\n",
    "#install gtdbtk v 2.5.2 (2.6.0 not working with the db unity has downloaded)\n",
    "conda create -p /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/.conda/envs/gtdbtk-2.5.2 -c conda-forge -c bioconda gtdbtk=2.5.2 python=3.12\n",
    "conda activate /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/.conda/envs/gtdbtk-2.5.2\n",
    "python --version #check that it is 3.12.12 and not 3.14\n",
    "\n",
    "conda env config vars set GTDBTK_DATA_PATH=\"/datasets/bio/gtdb/release226/226.0/auxillary_files/gtdbtk_package/full_package/release226\"\n",
    "conda deactivate #have to deactivate and reactivate env for the path change to take effect\n",
    "conda activate /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/.conda/envs/gtdbtk-2.5.2\n",
    "\n",
    "#check that the installation worked\n",
    "gtdbtk check_install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95541f23-5361-4cc9-8fb9-66cf58abafb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=150G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 48:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/slurms/slurm-gtdb_dlab-%j.out  # %j = job ID  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/.conda/envs/gtdbtk-2.5.2\n",
    "\n",
    "# Set parameters\n",
    "SAMPLENAME=\"dlab\"\n",
    "BINPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/binning/${SAMPLENAME}/Das_Tool/${SAMPLENAME}_DASTool_bins\"\n",
    "OUTDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/taxonomy/combined_data_MAGs/${SAMPLENAME}\"\n",
    "mkdir -p $OUTDIR\n",
    "\n",
    "#Run gtdb-tk\n",
    "gtdbtk classify_wf -x fa --genome_dir $BINPATH/ --out_dir $OUTDIR/gtdb_out\n",
    "\n",
    "#This will process all genomes in the directory <my_genomes> using both bacterial and archaeal marker sets and place the results in <output_dir>.\n",
    "#Genomes must be in FASTA format (gzip with the extension .gz is acceptable)\n",
    "\n",
    "# JOB-ID: 50572538 (this one worked: 50573688)\n",
    "# bash script file name: /nikea/COL/bash_scripts/Col_gtdb_taxonomy.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e4636e8-55d2-45f4-ab6f-2e22a1a81370",
   "metadata": {},
   "source": [
    "ugh all of my playing around with gtdbtk last week for cordap genome assembly and this still came back with\n",
    "EXCEPTION: ValueError\n",
    "  MESSAGE: \"__StageLogger\" object has no field \"version\"\n",
    "\n",
    "https://github.com/Ecogenomics/GTDBTk/issues/669 suggestion to downgrade from python 3.14 to python 3.12 so will try that"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f8b7178-7fe6-4172-ae3b-d6fcabd56f09",
   "metadata": {},
   "outputs": [],
   "source": [
    "salloc -c 6 -p cpu -t 04:00:00 \n",
    "module load conda/latest\n",
    "conda activate /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/.conda/envs/gtdbtk-2.5.2\n",
    "conda install -c conda-forge -c bioconda gtdbtk=2.5.2 python=3.12\n",
    "conda env config vars set GTDBTK_DATA_PATH=\"/datasets/bio/gtdb/release226/226.0/auxillary_files/gtdbtk_package/full_package/release226\"\n",
    "python --version #3.12.12\n",
    "gtdbtk test #(this comes back with the ValueError MESSAGE: \"__StageLogger\" object has no field \"version\")\n",
    "gtdbtk check_install #this completes without error\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a231e0bc-4b20-4c3f-a338-9e0dd7f77e2d",
   "metadata": {},
   "source": [
    "tried running Col_gtdb_taxonomy.sh again: 50573688\n",
    "I THINK THIS WORKED!!! so changing to python 3.12 was necessary - will make changes to above installation \n",
    "\n",
    "ran for the others: \\\n",
    "ofav: 50573739 \\\n",
    "pstr: 50573737 \\\n",
    "mcav: 50573747\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
