{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "79f3651e-a72d-4cba-bb3e-37170e14dd19",
   "metadata": {},
   "source": [
    "# Anvio workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0f07fb8-ff89-42d9-9a9b-69c868ff6feb",
   "metadata": {},
   "source": [
    "resource: https://merenlab.org/2016/06/22/anvio-tutorial-v2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b509c2b4-4952-4882-aa8d-f733419d01ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#INSTALLATION\n",
    "module load conda/latest\n",
    "conda activate anvio-8\n",
    "\n",
    "#additional set up\n",
    "anvi-setup-scg-taxonomy\n",
    "\n",
    "anvi-setup-ncbi-cogs --num-threads 11\n",
    "\n",
    "anvi-setup-kegg-data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e09b4c9-11c9-4b65-a0d7-0ba5e3d34480",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=100G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/anvio/mcav/slurm-anvio-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate anvio-8\n",
    "\n",
    "#Contig database from assembled genomes. stores information related to your sequences: positions of open reading frames, k-mer frequencies for each contig, functional and taxonomic annotation of genes, etc.\n",
    "#set parameters:\n",
    "SAMPLENAME=\"mcav\"\n",
    "CONTIGPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/mapping/${SAMPLENAME}\"\n",
    "CONTIGFILE=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "BAMPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/mapping/${SAMPLENAME}\"\n",
    "DBPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/anvio/${SAMPLENAME}\"\n",
    "mkdir -p \"$DBPATH\"\n",
    "OUTDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/anvio/${SAMPLENAME}/profiles\"\n",
    "mkdir -p \"$OUTDIR\"\n",
    "\n",
    "\n",
    "#from merged, fixed contig fasta file created in previous step..need for downstream analysis\n",
    "#default k-mer frequency is 4\n",
    "anvi-gen-contigs-database -f $CONTIGPATH/$CONTIGFILE --project-name $SAMPLENAME -o $DBPATH/$SAMPLENAME.contigs.db  \n",
    "\n",
    "#integrate HMMs into the database:\n",
    "anvi-run-hmms -c $DBPATH/$SAMPLENAME.contigs.db --num-threads 6\n",
    "\n",
    "#setup our COG database\n",
    "#anvi-setup-ncbi-cogs -T 11 --just-do-it\n",
    "\n",
    "#this runs NCBI COGs against your contigs.db, integrating gene functions\n",
    "anvi-run-ncbi-cogs -c $DBPATH/$SAMPLENAME.contigs.db -T 4\n",
    "\n",
    "#ADD KEGG-KOFAM\n",
    "anvi-run-kegg-kofams -c $DBPATH/$SAMPLENAME.contigs.db \\\n",
    "                     -T 4 #these are the threads that Anvi'O is allowed to use\n",
    "#ADD CONTIG STATS\n",
    "anvi-display-contigs-stats $DBPATH/$SAMPLENAME.contigs.db --report-as-text --as-markdown -o $DBPATH/anvio_stats.txt\n",
    "\n",
    "cd $BAMPATH\n",
    "#create sample profiles\n",
    "for f in $(cat 032024_mcav_sampleids.txt) \n",
    "do\n",
    "    echo \"processing $f file\" \n",
    "anvi-profile -c $DBPATH/$SAMPLENAME.contigs.db  \\\n",
    "            -i \"$f\".bam \\\n",
    "            --min-percent-identity 95 \\\n",
    "            --min-contig-length 1500 \\\n",
    "            --sample-name \"mcav\"_\"$f\" \\\n",
    "            --output-dir $OUTDIR/\"$f\"\n",
    "done\n",
    "#Will fail if output dir already exists. sample name can't start with an number, hence the mcav in front\n",
    "#keep parameters consistent in order to merge to larger profile \n",
    "echo \"Anvio profiling: All samples sucessfully completed!\"\n",
    "\n",
    "\n",
    "#merge single sample profiles to one profile\n",
    "anvi-merge $OUTDIR/*/PROFILE.db \\\n",
    "            -c $DBPATH/$SAMPLENAME.contigs.db \\\n",
    "            -o $OUTDIR/\"$SAMPLENAME\"_profile_merged\n",
    "\n",
    "echo \"Anvio profiling: All profiles merged\"\n",
    "conda deactivate\n",
    "\n",
    "# JOB-ID: 27300405, 27353332,27359919, 27374000\n",
    "# bash script file name: /nikea/COL/bash_scripts/Col_anvio.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d4b0f1b-69c6-45c9-bd92-beccf5dac132",
   "metadata": {},
   "source": [
    "failed at \"create sample profile step\". After lots of troubleshooting, got it to run. The output directory needs to be unique for each sample and not made ahead of time. Did piecemeal for mcav to get script running smoothly (that is why there are 4 job IDs). The whole script should run smoothly for the other species!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982ed36a-573a-4a58-b8fa-8cf58a81c348",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -c 24  # Number of Cores per Task\n",
    "#SBATCH --mem=50G  # Requested Memory\n",
    "#SBATCH -p cpu  # Partition\n",
    "#SBATCH -t 24:00:00  # Job time limit\n",
    "#SBATCH --mail-type=ALL\n",
    "#SBATCH -o /work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/anvio/mcav/slurm-anvio-%j.out  # %j = job ID\n",
    "\n",
    "module load conda/latest\n",
    "conda activate anvio-8\n",
    "\n",
    "#Contig database from assembled genomes. stores information related to your sequences: positions of open reading frames, k-mer frequencies for each contig, functional and taxonomic annotation of genes, etc.\n",
    "#set parameters:\n",
    "SAMPLENAME=\"ofav\"\n",
    "CONTIGPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/mapping/${SAMPLENAME}\"\n",
    "CONTIGFILE=\"${SAMPLENAME}.contigs-fixed.fsa\"\n",
    "BAMPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/mapping/${SAMPLENAME}\"\n",
    "DBPATH=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/anvio/${SAMPLENAME}\"\n",
    "mkdir -p \"$DBPATH\"\n",
    "OUTDIR=\"/work/pi_sarah_gignouxwolfsohn_uml_edu/nikea/COL/anvio/${SAMPLENAME}/profiles\"\n",
    "mkdir -p \"$OUTDIR\"\n",
    " \n",
    "\n",
    "#from merged, fixed contig fasta file created in previous step..need for downstream analysis\n",
    "#default k-mer frequency is 4\n",
    "anvi-gen-contigs-database -f $CONTIGPATH/$CONTIGFILE --project-name $SAMPLENAME -o $DBPATH/$SAMPLENAME.contigs.db  \n",
    "\n",
    "#integrate HMMs into the database:\n",
    "anvi-run-hmms -c $DBPATH/$SAMPLENAME.contigs.db --num-threads 6\n",
    "\n",
    "#this runs NCBI COGs against your contigs.db, integrating gene functions\n",
    "anvi-run-ncbi-cogs -c $DBPATH/$SAMPLENAME.contigs.db -T 4\n",
    "\n",
    "#ADD KEGG-KOFAM\n",
    "anvi-run-kegg-kofams -c $DBPATH/$SAMPLENAME.contigs.db \\\n",
    "                     -T 4 #these are the threads that Anvi'O is allowed to use\n",
    "#ADD CONTIG STATS\n",
    "anvi-display-contigs-stats $DBPATH/$SAMPLENAME.contigs.db --report-as-text --as-markdown -o $DBPATH/anvio_stats.txt\n",
    "\n",
    "cd $BAMPATH\n",
    "\n",
    "#create sample profiles\n",
    "#make sure sample list is in the folder w/ BAM files\n",
    "for f in $(cat 032024_ofav_sampleids.txt) \n",
    "do\n",
    "    echo \"processing $f file\" \n",
    "anvi-profile -c $DBPATH/$SAMPLENAME.contigs.db  \\\n",
    "            -i \"$f\".bam \\\n",
    "            --min-percent-identity 95 \\\n",
    "            --min-contig-length 1500 \\\n",
    "            --sample-name \"ofav\"_\"$f\" \\\n",
    "            --output-dir $OUTDIR/\"$f\"\n",
    "done\n",
    "#Will fail if output dir already exists. sample name can't start with an number, hence the species name in front\n",
    "#keep parameters consistent in order to merge to larger profile \n",
    "echo \"Anvio profiling: All samples sucessfully completed!\"\n",
    "\n",
    "#merge single sample profiles to one profile\n",
    "anvi-merge $OUTDIR/*/PROFILE.db \\\n",
    "            -c $DBPATH/$SAMPLENAME.contigs.db \\\n",
    "            -o $OUTDIR/\"$SAMPLENAME\"_profile_merged\n",
    "\n",
    "echo \"Anvio profiling: All profiles merged\"\n",
    "conda deactivate\n",
    "\n",
    "# JOB-ID: 27377112\n",
    "# bash script file name: /nikea/COL/bash_scripts/Col_anvio.sh"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
